File: module/build_cmds.py
import logging
import pathlib
import csv


from module import properties
from module import obi_constants
from module import toml_tools, files, app_config_tools
from copy import deepcopy
import re

default_app_config = properties.get_app_properties()



def add_build_cmds(target_tree, app_config=default_app_config):

  object_list = []

  for target_item in target_tree:
    for source_item in target_item['sources']:
      object_list.append(get_object_list(source_item['source'], app_config))
      source_item['cmds'] = get_source_build_cmds(source_item['source'], app_config)

  object_list = "\n".join(list(set(object_list)))
  logging.debug(f"Write object list {len(object_list)=} to {app_config['general']['deployment-object-list']}")
  files.writeText(object_list, app_config['general']['deployment-object-list'], write_empty_file=True)




def get_object_list(source, app_config=default_app_config):
  '''
  This is only needed by the deployment tool
  '''

  variable_dict = properties.get_source_properties(app_config, source)
  logging.debug(f"{source=}")
  logging.debug(f"{variable_dict=}")

  # Add list obj objects for deployment tool
  prod_lib = pathlib.Path(source).parts[0]
  #type=*pgm, attr=rpgle
  obj_type=source.split('.')[-1]
  obj_attr=source.split('.')[-2]
  deploy_obj_list = f"prod_obj|{prod_lib}|{variable_dict['TARGET_LIB']}|{variable_dict['OBJ_NAME']}|{obj_type}|{obj_attr}|{source}"
  
  return deploy_obj_list




def remove_unresolved_cmd_parameters(cmd: str) -> str:

  cmd = cmd.replace('ACTGRP($(ACTGRP))', '')
  cmd = cmd.replace('ACTGRP()', '')
  cmd = cmd.replace('BNDDIR($(INCLUDE_BNDDIR))', '')
  cmd = cmd.replace('BNDDIR()', '')
  cmd = cmd.replace('TGTRLS($(TGTRLS))', '')
  cmd = cmd.replace('TGTRLS()', '')
  cmd = cmd.replace('STGMDL($(STGMDL))', '')
  cmd = cmd.replace('STGMDL()', '')
  cmd = cmd.replace('TGTCCSID($(TGTCCSID))', '')
  cmd = cmd.replace('TGTCCSID()', '')
  cmd = cmd.replace('DBGVIEW($(DBGVIEW))', '')
  cmd = cmd.replace('DBGVIEW()', '')
  cmd = cmd.replace('INCDIR($(INCDIR_SQLRPGLE))', '')
  cmd = cmd.replace('INCDIR()', '')
  cmd = cmd.replace('INCDIR($(INCDIR_RPGLE))', '')
  cmd = cmd.replace('INCDIR()', '')

  return cmd



def get_source_build_cmds(source, app_config=default_app_config):

  logging.debug(f"Check source cmds for {source}")
  cmds = []
  sources_config=properties.get_config(obi_constants.OBIConstants.get("SOURCE_CONFIG_TOML"))
  source_config={}
  if source in sources_config:
    source_config = sources_config[source]


  src_suffixes = pathlib.Path(source).suffixes
  file_extensions = "".join(src_suffixes[-2:]).removeprefix('.')
  logging.debug(f"{file_extensions=}")
  logging.debug(f"{source_config=}")

  steps = app_config_tools.get_steps(source, app_config)
  #steps = app_config['global']['steps'].get(file_extensions, [])

  # Override steps by individual source config
  if 'steps' in source_config.keys() and len(source_config['steps']) > 0:
    steps = source_config['steps']
    logging.debug(f"{steps=}")

  # All properties for this source
  variable_dict = properties.get_source_properties(app_config, source)
  var_dict_tmp = {}
  logging.debug(f"{variable_dict=}")

  # Loop all steps of the source extension
  for step in steps:
    
    logging.debug(f"{step=}")

    if isinstance(step, str):
      cmd = get_cmd_from_step(step, source, variable_dict, app_config, source_config)
      logging.debug(f"1 {cmd=}")
      
    if isinstance(step, dict):
      var_dict_tmp = deepcopy(variable_dict)
      logging.debug(f"{var_dict_tmp=}")
      var_dict_tmp.update(step.get('properties', {}))
      
      cmd = step.get('cmd', None)
      if not cmd:
        cmd = get_cmd_from_step(step.get('step', None), source, var_dict_tmp, app_config, source_config)
        logging.debug(f"2 {cmd=}")
      
    cmd = replace_cmd_parameters(cmd, {**variable_dict, **var_dict_tmp})
    cmds.append({"cmd": cmd, "status": "new"})

  logging.debug(f"Added {len(cmds)} cmds for {source}")

  return cmds



def get_cmd_from_step(step: str, source: str, variable_dict: {}, app_config, source_config) -> str:
  
    logging.debug(f"csv.reader {step=}")
    r=csv.reader([step], quotechar='"', delimiter='.')  # step: e.g. global."compile-cmds"."sqlrpgle.srvpgm"
    step_list = next(r)                                 # --> ['global', 'compile-cmds', 'sqlrpgle.mod']
    #logging.debug(f"{step_list=}")

    #cmd = toml_tools.get_table_element(app_config, step_list)
    logging.debug(f"1: {step_list=}")
    cmd = toml_tools.get_table_element({**app_config, **source_config}, step_list)
    logging.debug(f"2: {cmd=}")
    
    # Find all words starting and ending with %
    percent_words = re.findall(r'%[\w\.]+%', cmd)
    logging.debug(f"Words starting and ending with %: {percent_words}")
    
    for word in percent_words:
        key = word.strip('%')
        r2=csv.reader([key], quotechar='"', delimiter='.')  # step: e.g. global."compile-cmds"."sqlrpgle.srvpgm"
        step_list2 = next(r2)                                 # --> ['global', 'compile-cmds', 'sqlrpgle.mod']
        logging.debug(f"Replace word {word} with key {key}")
        subcmd = toml_tools.get_table_element({**app_config, **source_config}, step_list2)
    
        if subcmd is not None:
          cmd = cmd.replace(word, subcmd)
          logging.debug(f"Replaced {word} with {subcmd} in cmd")

    variable_dict['SET_LIBL'] = properties.get_set_libl_cmd(app_config, variable_dict.get('LIBL', []), variable_dict['TARGET_LIB'])
    
    if cmd is None or cmd == '':
      raise Exception(f"Step '{step}' not found in '{obi_constants.OBIConstants.get("CONFIG_TOML")}' or '{obi_constants.OBIConstants.get("CONFIG_USER_TOML")}'")

    dspjoblog_cmd = app_config['global']['cmds'].get('dspjoblog', None)
    if dspjoblog_cmd is not None:
      joblog_sep = app_config['global']['cmds'].get('joblog-separator', "")
      cmd += dspjoblog_cmd.replace("$(joblog-separator)", joblog_sep)

    return cmd




def replace_cmd_parameters(cmd: str, variable_dict: {}) -> str:
  '''
  Replace all $(...) parameters in the cmd with the values from the variable_dict
  '''
  
  for k, v in variable_dict.items():
    if not isinstance(v, str) and not isinstance(v, int):
      continue
    cmd = cmd.replace(f"$({k})", str(v))

  cmd = remove_unresolved_cmd_parameters(cmd)

  return cmd



def order_builds(target_tree):
  '''
  Order the build commands by the order of the source files
  '''
  ordered_target_tree = {'timestamp': target_tree['timestamp'], 'compiles': []}
  
  for compiles in target_tree['compiles']:
    
    level_list = {'level': compiles['level'], 'sources': []}
    
    for source_entry in compiles['sources']:
      
      if source_entry['source'].split('.')[-1] == 'file':
        level_list['sources'].insert(0, source_entry)
        continue
      
      level_list['sources'].append(source_entry)
      
    ordered_target_tree['compiles'].append(level_list)
  
  return ordered_target_tree

File: src/module/build_cmds.ts
// TypeScript conversion of module/build_cmds.py

import * as path from 'path';
import * as properties from './properties';
import * as files from './files';
import * as appConfigTools from './app_config_tools';
import * as tomlTools from './toml_tools';
import { OBIConstants } from './obi_constants';

interface SourceItem {
    source: string;
    cmds: CommandItem[];
}

interface CommandItem {
    cmd: string;
    status: string;
}

interface LevelItem {
    level: number;
    sources: SourceItem[];
}

export function addBuildCmds(targetTree: LevelItem[], appConfig: any): void {
    const objectList: string[] = [];

    for (const targetItem of targetTree) {
        for (const sourceItem of targetItem.sources) {
            objectList.push(getObjectList(sourceItem.source, appConfig));
            sourceItem.cmds = getSourceBuildCmds(sourceItem.source, appConfig);
        }
    }

    const objectListStr = Array.from(new Set(objectList)).join('\n');
    console.log(`Write object list length=${objectListStr.length} to ${appConfig['general']['deployment-object-list']}`);
    files.writeText(objectListStr, appConfig['general']['deployment-object-list'], true);
}

function getObjectList(source: string, appConfig: any): string {
    const variableDict = properties.getSourceProperties(appConfig, source);
    console.log(`source=${source}`);
    console.log(`variableDict=${JSON.stringify(variableDict)}`);

    const prodLib = source.split('/')[0];
    const objType = source.split('.').pop() || '';
    const objAttr = source.split('.').slice(-2, -1)[0] || '';
    const deployObjList = `prod_obj|${prodLib}|${variableDict['TARGET_LIB']}|${variableDict['OBJ_NAME']}|${objType}|${objAttr}|${source}`;
    
    return deployObjList;
}

function removeUnresolvedCmdParameters(cmd: string): string {
    cmd = cmd.replace(/ACTGRP\(\$\(ACTGRP\)\)/g, '');
    cmd = cmd.replace(/ACTGRP\(\)/g, '');
    cmd = cmd.replace(/BNDDIR\(\$\(INCLUDE_BNDDIR\)\)/g, '');
    cmd = cmd.replace(/BNDDIR\(\)/g, '');
    cmd = cmd.replace(/TGTRLS\(\$\(TGTRLS\)\)/g, '');
    cmd = cmd.replace(/TGTRLS\(\)/g, '');
    cmd = cmd.replace(/STGMDL\(\$\(STGMDL\)\)/g, '');
    cmd = cmd.replace(/STGMDL\(\)/g, '');
    cmd = cmd.replace(/TGTCCSID\(\$\(TGTCCSID\)\)/g, '');
    cmd = cmd.replace(/TGTCCSID\(\)/g, '');
    cmd = cmd.replace(/DBGVIEW\(\$\(DBGVIEW\)\)/g, '');
    cmd = cmd.replace(/DBGVIEW\(\)/g, '');
    cmd = cmd.replace(/INCDIR\(\$\(INCDIR_SQLRPGLE\)\)/g, '');
    cmd = cmd.replace(/INCDIR\(\)/g, '');
    cmd = cmd.replace(/INCDIR\(\$\(INCDIR_RPGLE\)\)/g, '');
    return cmd;
}

export function getSourceBuildCmds(source: string, appConfig: any): CommandItem[] {
    console.log(`Check source cmds for ${source}`);
    const cmds: CommandItem[] = [];
    const sourcesConfig = properties.getConfig(OBIConstants.get("SOURCE_CONFIG_TOML") || '.obi/source-config.toml');
    const sourceConfig = source in sourcesConfig ? sourcesConfig[source] : {};

    const srcSuffixes = path.extname(source);
    const fileExtensions = srcSuffixes.replace(/^\./, '');
    console.log(`fileExtensions=${fileExtensions}`);
    console.log(`sourceConfig=${JSON.stringify(sourceConfig)}`);

    let steps = appConfigTools.getSteps(source, appConfig);

    // Override steps by individual source config
    if ('steps' in sourceConfig && sourceConfig['steps'].length > 0) {
        steps = sourceConfig['steps'];
        console.log(`steps=${JSON.stringify(steps)}`);
    }

    const variableDict = properties.getSourceProperties(appConfig, source);
    let varDictTmp: any = {};
    console.log(`variableDict=${JSON.stringify(variableDict)}`);

    for (const step of steps) {
        console.log(`step=${JSON.stringify(step)}`);
        if (typeof step === 'string' && step.trim() === '') {
            continue;
        }

        let cmd: string;
        
        if (typeof step === 'string') {
            cmd = getCmdFromStep(step, source, variableDict, appConfig, sourceConfig);
            console.log(`1 cmd=${cmd}`);
        } else if (typeof step === 'object') {
            varDictTmp = JSON.parse(JSON.stringify(variableDict)); // Deep copy
            console.log(`varDictTmp=${JSON.stringify(varDictTmp)}`);
            Object.assign(varDictTmp, step['properties'] || {});
            
            cmd = step['cmd'] || '';
            if (!cmd) {
                cmd = getCmdFromStep(step['step'] || '', source, varDictTmp, appConfig, sourceConfig);
                console.log(`2 cmd=${cmd}`);
            }
        } else {
            continue;
        }
        
        cmd = replaceCmdParameters(cmd, { ...variableDict, ...varDictTmp });
        cmds.push({ cmd, status: "new" });
    }

    console.log(`Added ${cmds.length} cmds for ${source}`);

    return cmds;
}

function getCmdFromStep(step: string, source: string, variableDict: any, appConfig: any, sourceConfig: any): string {
    let cmd = resolveCmdid({ ...appConfig, ...sourceConfig }, step);
    console.log(`2: cmd=${cmd}`);
    
    // Find all words starting and ending with %
    const percentWords = cmd.match(/%[\w\.]+%/g) || [];
    console.log(`Words starting and ending with %: ${JSON.stringify(percentWords)}`);
    
    for (const word of percentWords) {
        const key = word.replace(/%/g, '');
        const subcmd = resolveCmdid({ ...appConfig, ...sourceConfig }, key);
        cmd = cmd.replace(word, subcmd);
        console.log(`Replaced ${word} with ${subcmd} in cmd`);
    }

    variableDict['SET_LIBL'] = properties.getSetLiblCmd(appConfig, variableDict['LIBL'] || [], variableDict['TARGET_LIB']);
    
    if (!cmd || cmd === '') {
        throw new Error(`Step '${step}' not found in config`);
    }

    const dspjoblogCmd = appConfig['global']['cmds']['dspjoblog'] || null;
    if (dspjoblogCmd !== null) {
        const joblogSep = appConfig['global']['cmds']['joblog-separator'] || "";
        cmd += dspjoblogCmd.replace("$(joblog-separator)", joblogSep);
    }

    return cmd;
}

function resolveCmdid(config: any, cmdid: string): string {
    // Simple CSV parsing for cmdid like "global.compile-cmds.sqlrpgle.mod"
    const cmdidList = cmdid.split('.');
    console.log(`cmdidList=${JSON.stringify(cmdidList)}`);

    const cmd = tomlTools.getTableElement(config, cmdidList);
    
    if (!cmd || cmd === '') {
        throw new Error(`CmdID '${cmdid}' not found in config`);
    }

    return cmd;
}

function replaceCmdParameters(cmd: string, variableDict: any): string {
    for (const [k, v] of Object.entries(variableDict)) {
        if (typeof v !== 'string' && typeof v !== 'number') {
            continue;
        }
        cmd = cmd.replace(new RegExp(`\\$\\(${k}\\)`, 'g'), String(v));
    }

    cmd = removeUnresolvedCmdParameters(cmd);

    return cmd;
}

export function orderBuilds(targetTree: { timestamp: string; compiles: LevelItem[] }): { timestamp: string; compiles: LevelItem[] } {
    const orderedTargetTree: { timestamp: string; compiles: LevelItem[] } = {
        timestamp: targetTree.timestamp,
        compiles: []
    };
    
    for (const compiles of targetTree.compiles) {
        const levelList: LevelItem = { level: compiles.level, sources: [] };
        
        for (const sourceEntry of compiles.sources) {
            if (sourceEntry.source.split('.').pop() === 'file') {
                levelList.sources.unshift(sourceEntry);
                continue;
            }
            
            levelList.sources.push(sourceEntry);
        }
        
        orderedTargetTree.compiles.push(levelList);
    }
    
    return orderedTargetTree;
}
